Implementation of a Limited QASM3 Exporter except controlflow and Classical Variables in Rust

to24toro opened the pull request:
Fixes #13145 and Continue with #13666

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->
### Summary
This PR introduces and integrates several key modules for a Rust-based quantum computing framework, laying the foundation for exporting circuits in QASM 3.0 format. The modules added are:
**1. printer.rs:**
  • Implements functionality for formatted output generation of AST or other hierarchical structures.
**2. symbols.rs:**
  • Manages a symbol table used for mapping variable names to their corresponding entities.
  • Handling bit info, register info, and gate info in the symbol table.
  • Provides helper methods for creating, binding, and looking up symbols with robust error handling.
**3. ast.rs:**
  • Defines the AST structure for representing quantum programs.
  • Contains data structures such as nodes, expressions, and declarations that model the core components of quantum circuit descriptions.
**4. exporter.rs:**
  • Implements a builder for generating QASM 3.0 programs.
  • Integrates with Python-based tools (via pyo3) to extract circuit data and parameters.

### Details and comments
#### Current Limitations

The current implementation has limited QASM 3.0 output functionality and relies on partial Python integration. The following features are not yet supported:
  • Control Flow Constructs: While QASM 3.0 supports control flow, such as loops and conditionals, these are not yet implemented.
  • Classical Variables: Handling classical variables is not included in the current architecture.
  • Not standard Gates: Generating QASM for not standard gates and mapping them to Qiskit’s standard gates is pending.

These tasks require further discussion to determine whether they should be included in the scope of Qiskit Issue #13145. 

#### What is Pending
**1. Support for:**
  • Control flow constructs like loops, conditionals, and subroutines.
  • Advanced QASM 3.0 features, including classical variable.

#### Difference from previous PR(#13666 )
• Support for Registers.
• Support for all the standard gates defined by qiskit.
• Support for custom gates.
• Add tests in test_exporter.py.

We can export the below circuit.

```python
from qiskit import qasm3
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
from qiskit.circuit import Clbit, Qubit
from qiskit.circuit.library import XXPlusYYGate
qr = QuantumRegister(2,"a")
cr = ClassicalRegister(1)
qc = QuantumCircuit(2)
qc.add_bits([Clbit()])
qc.add_register(qr)
qc.add_bits([Qubit()])
qc.add_register(cr)
qc.append(XXPlusYYGate(0.4,0.12),[0,1])
qc.measure_all()
print(qasm3.dumps_experimental(qc))
```

Output is here:

```
OPENQASM 3.0;
include "stdgates.inc";
gate sxdg _gate_q_0 {
  s _gate_q_0;
  h _gate_q_0;
  s _gate_q_0;
}
gate xx_plus_yy(_gate_p_0, _gate_p_1) _gate_q_0, _gate_q_1 {
  rz(_gate_p_1) _gate_q_0;
  rz(-1.5707963267948966) _gate_q_1;
  sx _gate_q_1;
  rz(1.5707963267948966) _gate_q_1;
  s _gate_q_0;
  cx _gate_q_1, _gate_q_0;
  ry(-0.5*_gate_p_0) _gate_q_1;
  ry(-0.5*_gate_p_0) _gate_q_0;
  cx _gate_q_1, _gate_q_0;
  sdg _gate_q_0;
  rz(-1.5707963267948966) _gate_q_1;
  sxdg _gate_q_1;
  rz(1.5707963267948966) _gate_q_1;
  rz(-1.0*_gate_p_1) _gate_q_0;
}
bit _bit0;
bit[1] c1;
bit[5] meas;
qubit _qubit4;
qubit[2] q;
qubit[2] a;
xx_plus_yy(0.4, 0.12) q[0], q[1];
barrier q[0], q[1], a[0], a[1], _qubit4;
meas[0] = measure q[0];
meas[1] = measure q[1];
meas[2] = measure a[0];
meas[3] = measure a[1];
meas[4] = measure _qubit4;
```


```python
import numpy as np
import warnings

from qiskit import qasm3
from qiskit.exceptions import ExperimentalWarning
from qiskit.circuit.library import RealAmplitudes

warnings.filterwarnings("ignore", category=ExperimentalWarning)

# This is a little more than 5k gates.
base = RealAmplitudes(27, reps=100, flatten=True)
base = base.assign_parameters(np.random.rand(base.num_parameters))

print("Current OpenQASM 3 time")
%timeit qasm3.dumps(base)
print("New OpenQASM 3 time")
%timeit qasm3.dumps_experimental(base)
```

```
Current OpenQASM 3 time
45.6 ms ± 282 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
New OpenQASM 3 time
4.33 ms ± 165 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

to24toro committed:
draft for qasm3exporter

to24toro committed:
refactoring exporter and ast

to24toro committed:
Merge remote-tracking branch 'origin/main' into feature/qasm3exporter

to24toro committed:
wip: simplify symbol table

to24toro committed:
Remove unused struct

to24toro committed:
Add #[allow(dead_code)] to ast.rs

to24toro committed:
Remove scopes from symbol_table because we might not support controlflow and scope switching

to24toro committed:
Remove qubit_map, clbit_map, and some functions for handling controlflow beucase we might not support controlflow

to24toro committed:
Add tests and format

to24toro committed:
format

to24toro committed:
Merge branch 'main' into feature/qasm3exporter

to24toro committed:
Merge branch 'main' into feature/qasm3exporter

to24toro committed:
Merge branch 'main' into feature/qasm3exporter

to24toro committed:
Merge branch 'main' into feature/qasm3exporter

to24toro committed:
Added registers to BitLocations

to24toro committed:
Edited ast.rs

to24toro committed:
Exporter.rs for standard gates

to24toro committed:
Fixed exporter.py

to24toro committed:
printer.rs

to24toro committed:
Add test

to24toro committed:
format

to24toro committed:
python format

to24toro committed:
Fix dumps from python

to24toro committed:
refactoring exporter.rs

to24toro committed:
Fix printer.py

to24toro committed:
Reorder each functions

to24toro committed:
Merge branch 'main' into feature/qasm3exporter

jakelishman reviewed (commented):
Thanks for this.  I haven't looked through entirely completely yet, but some top-level comments I had from a quick play:

1. I think you might have run your performance numbers with a debug build?  I see rather better performance than you did.
2. The order that gate definitions (like `xx_plus_yy` and `sxdg` in the example) is non-deterministic - it can change if you restart `python`.  We want it to be deterministic, but more importantly, the order is very important - in the example, `sxdg` _must_ be defined before `xx_plus_yy`, because `xx_plus_yy` uses `sxdg`.
3. The AST doesn't interact with any sort of symbol table, and it takes most things by ownership, including (and most importantly) `Identifier`.  So when we're outputting a circuit with many many gates, we're first allocating lots and lots of `String` instances with the same value, then copying them over into the output.  It will likely be much faster to have the AST be defined in reference to a symbol table, so the string representations of identifiers are stored separately, and the AST only stores an integer index into the table.  You can do this completely separately if you like, or you might find an `Interner<str>` from `qiskit_circuit::interner` to be a quick way to get something _good enough_.

to24toro committed:
Replace hashmap with indexmap in using gates in symboltable

to24toro commented:
@jakelishman 
Thank you for your comments.

1. I forgot to execute running with a release mode. I updated the performance metrics.
2. I changed to use `Indexmap` instead of `hashmap` at https://github.com/Qiskit/qiskit/pull/14226/commits/fbcf28a00dab1c2fe095ca81ce5f69afcb7a5e9b .
3. As you mentioned, I also believe it can be made faster, and I have confirmed it — it was about 12× faster on my local PC by avoiding string copying when creating `Identifier`.  But I want to separate a PR because I am not confident whether we can safely use this symbol table architecture that includes `qiskit_circuit::interner` when introducing controlflow statements, that is, using `new_scope`. I am afraid that it will be difficult to correctly manage the appropriate Interner across different scopes.

to24toro committed:
Merge branch 'main' into feature/qasm3exporter

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into feature/qasm3exporter

A syntactic (but not logical) merge conflict in `qiskit_circuit::bit`
was trivial, but a logical merge conflict caused by the PyO3 0.25 update
needed some actual code changes.

jakelishman commented on crates/qasm3/src/ast.rs:
```diff
@@ -0,0 +1,446 @@
+    pub subscript: Box<Expression>,
+}
+
+#[derive(Debug, Clone)]
+pub struct IntegerLiteral(pub(crate) i32);
```
There are likely problems ahead of us if we're limiting integers to 32 bits (registers are in theory arbitrary sized), but we can look at changing the internal representation of that in the future.

jakelishman commented on crates/qasm3/src/ast.rs:
```diff
@@ -0,0 +1,446 @@
+#[derive(Debug, Clone)]
+pub struct BooleanLiteral(pub(crate) bool);
+
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub struct BitstringLiteral {
+    pub value: String,
+    pub width: u32,
+}
```
A bitstring literal can always be represented by a (big) integer, so strictly some `u64` or unsigned `BigInteger` would be more efficient here, but again, we can look at that later.

jakelishman commented on crates/qasm3/src/ast.rs:
```diff
@@ -0,0 +1,446 @@
+    pub width: u32,
+}
+
+#[derive(Debug, Clone)]
+pub struct DurationLiteral {
+    pub value: f64,
+    pub unit: DurationUnit,
+}
+
+#[derive(Debug, Clone)]
+pub enum DurationUnit {
+    Nanosecond,
+    Microsecond,
+    Millisecond,
+    Second,
+    Sample,
+}
```
Since this PR was written, we gained a more Rust-native way to represent durations in `qiskit_circuit`, so in the future we can swap to that.

jakelishman commented on crates/qasm3/src/printer.rs:
```diff
@@ -0,0 +1,631 @@
+        }
+    }
+
+    fn start_line(&mut self) {
+        write!(self.stream, "{}", self.indent.repeat(self.current_indent)).unwrap();
+    }
```
All the way through the printer: panicking in Rust is generally supposed to be for un-recoverable errors, particularly in a library, like Qiskit.  We should ideally return a `::std::io::Result<()>` from each of these functions, and bubble up the failure properly to Python space.

jakelishman commented on crates/qasm3/src/printer.rs:
```diff
@@ -0,0 +1,631 @@
+    }
+
+    fn visit_header(&mut self, node: &Header) {
+        if let Some(version) = &node.version {
+            self.visit(&Node::Version(version))
+        };
+        for include in node.includes.iter() {
+            self.visit(&Node::Include(include));
+        }
```
Very minor: wrapping a `Version` object in a `Node::Version` in order to dispatch it back down to `Version` just ends up being a bit of a waste of time (though I'd imagine the compiler will actually optimise it out here) - there's no particular reason not to just call `visit_version` directly (and so on for other similar examples).

jakelishman commented on crates/qasm3/src/printer.rs:
```diff
@@ -0,0 +1,631 @@
+        }
+    }
+
+    fn visit_include(&mut self, node: &Include) {
+        self.write_statement(&format!("include \"{}\"", node.filename));
+    }
+
+    fn visit_version(&mut self, node: &Version) {
+        self.write_statement(&format!("OPENQASM {}", node.version_number));
+    }
```
When we're optimising for performance in the future: the `write_statement` method here is promoting inefficient code, because we're heap allocating a `String`, just to write it out again with the indent prepended and a semicolon appended.  The `write!` macros accepts formatting arguments, so generally we want to design our internal APIs and helper functions to be able to use that, not `format!`.

jakelishman commented on crates/qasm3/src/printer.rs:
```diff
@@ -0,0 +1,631 @@
+            },
+            Expression::IntegerLiteral(expression) => self.visit_integer_literal(expression),
+            Expression::BooleanLiteral(expression) => self.visit_boolean_literal(expression),
+            Expression::BitstringLiteral(_) => {
+                panic!("BasicPrinter: BitStringLiteral has not been supported yet.")
+            }
```
This is another type of thing that should be a Python-space exception, not a Rust panic.

jakelishman commented on crates/qasm3/src/ast.rs:
```diff
@@ -0,0 +1,446 @@
+    ShiftRight,
+}
+
+impl Display for BinaryOp {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let op_str = match self {
+            BinaryOp::BitAnd => "&",
+            BinaryOp::BitOr => "|",
+            BinaryOp::BitXor => "^",
+            BinaryOp::LogicAnd => "&&",
+            BinaryOp::LogicOr => "||",
+            BinaryOp::Less => "<",
+            BinaryOp::LessEqual => "<=",
+            BinaryOp::Greater => ">",
+            BinaryOp::GreaterEqual => ">=",
+            BinaryOp::Equal => "==",
+            BinaryOp::NotEqual => "!=",
+            BinaryOp::ShiftLeft => "<<",
+            BinaryOp::ShiftRight => ">>",
+        };
+        write!(f, "{}", op_str)
+    }
+}
```
Potentially this might have been neater/more extensible to make the `match self {}` block a separate method called `as_str(&self) -> &'a static str` or something like that, and call it. It feels slightly odd to me to implement `Display`, but I think it's just personal preference, and this is fine.

jakelishman commented on crates/qasm3/src/ast.rs:
```diff
@@ -0,0 +1,446 @@
+    BitArray(BitArray),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Float {
+    Half = 16,
+    Single = 32,
+    Double = 64,
+    Quad = 128,
+    Oct = 256,
+}
```
Setting the discriminant here is probably not really what you intended, because it has implications on how the compiler may optimise them - using 256 as a value means (I think) it's required to use at least a `u16` to do it now.  You could also have a `width(&self) -> usize` method that returns the value.

If you _do_ do this, and you want to access the discriminant of these simple unit variants, you can use an `as` cast - `Float::Half as u16` would resolve to `16`.

jakelishman commented on crates/qasm3/src/ast.rs:
```diff
@@ -0,0 +1,446 @@
+    pub qubits: Vec<IdentifierOrSubscripted>,
+}
+
+#[derive(Debug, Clone)]
+pub struct QuantumMeasurementAssignment {
+    pub identifier: IdentifierOrSubscripted,
+    pub quantum_measurement: QuantumMeasurement,
+}
```
Qiskit doesn't actually support measuring a qubit and throwing away the result, so the `QuantumMeasurement` internal object here is probably a bit unnecessary, given it's just `Vec<IdentifierOrSubscripted>`.

jakelishman commented on crates/qasm3/src/ast.rs:
```diff
@@ -0,0 +1,446 @@
+    pub step: Option<Box<Expression>>,
+}
+
+#[derive(Debug, Clone)]
+pub enum IdentifierOrSubscripted {
+    Identifier(Identifier),
+    Subscripted(SubscriptedIdentifier),
+}
+
+#[derive(Debug, Clone)]
+pub struct Identifier {
+    pub string: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct SubscriptedIdentifier {
+    pub string: String,
+    pub subscript: Box<Expression>,
+}
```
We probably could have got rid of `IdentifierOrSubscripted`, and just like `SubscriptedIdentifier` handle it with `subscript: Option<Box<Expression>>`; it's mostly a convenience object that's a special case of an `Expression::Index` anyway.

jakelishman commented on crates/qasm3/src/ast.rs:
```diff
@@ -0,0 +1,446 @@
+#[derive(Debug, Clone)]
+pub struct Continue {}
+
+#[derive(Debug, Hash, Eq, PartialEq)]
+pub enum OP<'a> {
+    UnaryOp(&'a UnaryOp),
+    BinaryOp(&'a BinaryOp),
+}
```
If the purpose of this is just to be the wrapper type to allow the `binding_power` thing to work, we potentially could have left it be, and had static methods that returned the binding power of the two types separately; it doesn't really need to be a dynamic hashmap that both objects are in.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+    static ref _VALID_HARDWARE_QUBIT: Regex = Regex::new(r"\$\d+").unwrap();
+}
+
+#[derive(Error, Debug)]
+pub enum QASM3ExporterError {
+    #[error("Error: {0}")]
+    Error(String),
+    #[error("PyError: {0}")]
+    PyErr(PyErr),
+}
+
+impl From<PyErr> for QASM3ExporterError {
+    fn from(err: PyErr) -> Self {
+        QASM3ExporterError::PyErr(err)
+    }
+}
```
Typically the point of `thiserror` is so you can have your error enums be structured, and construct the error message lazily _if_ needed.  Rust programs shouldn't generally assume that a `Result` being in an error state is always irrecoverable, and they typically want structured data, especially if some are recoverable / carry useful information.

That said, most of the failure modes here are either from Python (in which case we have to use the type-erased `PyErr`), or are IO-related, which are typically handled opaquely.  This is fine for now, but in the future the logic of formatting the error message should probably live here.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+enum BitType {
+    ShareableQubit(ShareableQubit),
+    ShareableClbit(ShareableClbit),
+}
+
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+enum RegisterType {
+    QuantumRegister(QuantumRegister),
+    ClassicalRegister(ClassicalRegister),
+}
```
`ShareableQubit` (and clbit) is not really the correct type here for performance - we should always be able to resolve simply in terms of qubit indices using the `Interner`s that exist in the `CircuitData` objects.  Using the Python-space `Bit` objects was convenient in Python space when dealing with control flow, but it's not performant in Rust; rather than using hashmaps for these (and particularly _combining_ them into a hashmap), it'd be more performant for the symbol table to do the lookup by `Qubit` index into a `&[SubscriptedIdentifier]`, at least.

Really, we might well be able to propagate that all the way through to the printer, for performance, such as having the AST use the `qiskit_circuit::Qubit` type in all places that can only be a qubit (gate calls, etc, since Qiskit doesn't _store_ broadcasting), including resolving them to the outer circuit within control-flow scopes, and then giving the formatter access to the symbol table that stores the `&[SubscriptedIdentifier]` for the qubits and clbits.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+    ClassicalRegister(ClassicalRegister),
+}
+
+impl RegisterType {
+    fn name(&self) -> &str {
+        match self {
+            RegisterType::QuantumRegister(q) => q.name(),
+            RegisterType::ClassicalRegister(c) => c.name(),
+        }
+    }
+
+    fn bits(&self) -> Vec<BitType> {
+        match self {
+            RegisterType::QuantumRegister(quantum_register) => quantum_register
+                .bits()
+                .map(BitType::ShareableQubit)
+                .collect(),
+            RegisterType::ClassicalRegister(classical_register) => classical_register
+                .bits()
+                .map(BitType::ShareableClbit)
+                .collect(),
+        }
+    }
+}
```
I'm not certain that erasing the types here by adding a "wrapper" enum object is helping much here - is it providing a significant benefit over storing the lookups separately?

Combining tables of different types in the symbol table helps ensure we don't have naming collisions, but we only really get the benefit of that if _all_ symbols are put together.  Since they're not, we still have to check multiple tables, so I'm not sure that partial type erasure is winning us much.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+}
+
+struct BuildScope {
+    circuit_data: CircuitData,
+    bit_map: HashMap<BitType, BitType>,
+}
+
+impl BuildScope {
+    fn new(
+        circuit_data: &CircuitData,
+        qubits: &[ShareableQubit],
+        clbits: &[ShareableClbit],
+    ) -> Self {
+        let mut bit_map: HashMap<BitType, BitType> = HashMap::new();
+        for q in qubits.iter() {
+            bit_map.insert(
+                BitType::ShareableQubit(q.clone()),
+                BitType::ShareableQubit(q.clone()),
+            );
+        }
+
+        for c in clbits.iter() {
+            bit_map.insert(
+                BitType::ShareableClbit(c.clone()),
+                BitType::ShareableClbit(c.clone()),
+            );
+        }
+        Self {
+            circuit_data: circuit_data.clone(),
+            bit_map,
+        }
+    }
+
+    fn with_mappings(circuit_data: CircuitData, bit_map: HashMap<BitType, BitType>) -> Self {
+        Self {
+            circuit_data,
+            bit_map,
+        }
+    }
+}
```
This is another place where we could make use of the fact that qubit and clbit indices are contiguous, and that Rust preferentially deals with these indices, rather than `ShareableQubit`.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+    }
+}
+
+struct SymbolTable {
+    symbols: Vec<HashMap<String, Identifier>>,
+    bitinfo: Vec<HashMap<BitType, IdentifierOrSubscripted>>,
+    reginfo: Vec<HashMap<RegisterType, IdentifierOrSubscripted>>,
+    gates: IndexMap<String, QuantumGateDefinition>,
+    stdgates: HashSet<String>,
+    _counter: Counter,
+}
```
A few points on design for the future:
- the concept of the stacked scopes within the symbol table feels a bit awkward to be represented as three separate `Vec`s; it might be a bit cleaner to just have one, and have an intermediate `LocalSymbols` struct to hold the three separate arrays - that makes it impossible to even represent the (invalid) state that the vecs don't have the same length
- I think the distinction between `gates` and `stdgates` is meant to be "gates that need a definition" and "gates that don't need a definition"?  It might be clearer to name things like that.
- I'm not certain that tracking the `QuantumGateDefinition` AST node in the `SymbolTable` is the right separation of concerns - similar to how `QASM3Builder::global_io_decls` is tracking state for IO declarations.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+        }
+    }
+
+    fn bind(&mut self, name: &str) -> ExporterResult<()> {
+        if let Some(symbols) = self.symbols.last() {
+            if !symbols.contains_key(name) {
+                self.bind_no_check(name);
+            }
+        } else {
+            return Err(QASM3ExporterError::Error(format!(
+                "Symbol table is empty, cannot bind '{}'",
+                name
+            )));
+        }
+
+        Ok(())
+    }
+
+    fn bind_no_check(&mut self, name: &str) {
+        let id = Identifier {
+            string: name.to_string(),
+        };
+        if let Some(last) = self.symbols.last_mut() {
+            last.insert(name.to_string(), id);
+        }
+    }
```
Minor naming: typically if there's two variants of a function, one returns `Result` and one panics instead, it Rust convention to call the `Result` one `try_bind` and the panicking one `bind`.

Actually in this case, though, there should be no need for the `Result` from `bind`: that error represents a non-actionable internal logic error.  If we hit that path, the internal logic of the exporter _must_ be broken, which is a suitable condition for panicking.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+            self.allow_aliasing,
+        );
+        match builder.build_program() {
+            Ok(program) => {
+                let mut output = String::new();
+                BasicPrinter::new(&mut output, self.indent.to_string(), false)
+                    .visit(&Node::Program(&program));
+                Ok(output)
+            }
+            Err(e) => Err(QASM3ExporterError::Error(e.to_string())),
+        }
+    }
+
+    pub fn dump<W: Write>(
+        &self,
+        circuit_data: &CircuitData,
+        islayout: bool,
+        writer: &mut W,
+    ) -> ExporterResult<()> {
+        let mut builder = QASM3Builder::new(
+            circuit_data,
+            islayout,
+            self.includes.clone(),
+            self.basis_gates.clone(),
+            self.disable_constants,
+            self.allow_aliasing,
+        );
+
+        match builder.build_program() {
+            Ok(program) => {
+                let mut output = String::new();
+                let mut printer = BasicPrinter::new(&mut output, self.indent.to_string(), false);
+                printer.visit(&Node::Program(&program));
+                drop(printer);
+                let _ = writer.write_all(output.as_bytes());
+                Ok(())
+            }
+            Err(e) => Err(QASM3ExporterError::Error(e.to_string())),
+        }
+    }
```
If we need both of these, one of them should be defined in terms of the other, most likely `dumps` in terms of `dump`.

Currently `dump` is offering no performance benefit over a user manually creating the string and calling `writer.write_all(output.as_bytes())` themselves, but the API makes it look like it does.  Imo, in Rust, it's generally more idiomatic to make expensive compound operations look expensive by making the user do them.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+}
+
+pub struct QASM3Builder {
+    _builtin_instr: HashSet<&'static str>,
+    loose_bit_prefix: &'static str,
+    loose_qubit_prefix: &'static str,
+    _gate_param_prefix: &'static str,
+    _gate_qubit_prefix: &'static str,
```
If these are necessarily `&'static str`, we most likely don't need to store references to them, but can just pull them from the outer scope.

jakelishman commented on crates/qasm3/src/exporter.rs:
```diff
@@ -0,0 +1,1412 @@
+
+type ExporterResult<T> = Result<T, QASM3ExporterError>;
+
+// These are the prefixes used for the loose qubit and bit names.
+lazy_static! {
+    static ref BIT_PREFIX: &'static str = "_bit";
+}
+lazy_static! {
+    static ref QUBIT_PREFIX: &'static str = "_qubit";
+}
+
+// These are the prefixes used for the gate parameters and qubits.
+lazy_static! {
+    static ref GATE_PARAM_PREFIX: &'static str = "_gate_p";
+}
+lazy_static! {
+    static ref GATE_QUBIT_PREFIX: &'static str = "_gate_q";
+}
```
These ones don't need to be lazy - `const BIT_PREFIX: &str = "_bit";` would work fine.

jakelishman commented on crates/qasm3/src/lib.rs:
```diff
@@ -150,11 +154,147 @@ pub fn load(
     loads(py, source, custom_gates, include_path)
 }
 
+#[derive(Debug, Clone)]
+struct DumpOptions {
+    includes: Vec<String>,
+    basis_gates: Vec<String>,
+    disable_constants: bool,
+    allow_aliasing: bool,
+    indent: String,
+}
+
+impl Default for DumpOptions {
+    fn default() -> Self {
+        Self {
+            includes: vec!["stdgates.inc".to_string()],
+            basis_gates: vec![],
+            disable_constants: true,
+            allow_aliasing: false,
+            indent: "  ".to_string(),
+        }
+    }
+}
```
If these are just the options that go into making an `Exporter`, would it have made sense to skip the intermediate structure and go straight to having constructors of `Exporter`?  `Exporter` itself is just a configuration definition for `QASM3Builder`, and this is a configuration definition for `Exporter`.

jakelishman commented on crates/qasm3/src/lib.rs:
```diff
@@ -150,11 +154,147 @@ pub fn load(
+}
+
+#[pyfunction]
+#[pyo3(signature = (circuit, /, kwargs=None))]
+pub fn dumps(
+    _py: Python,
+    circuit: &Bound<PyAny>,
+    kwargs: Option<&Bound<PyDict>>,
+) -> PyResult<String> {
```
Making a fake `kwargs` dictionary like this is a bit unusual for a Rust interface.  You can specify keyword args properly using the `pyo3(signature = ...)` derive macro, like you have, including specifying defaults.  But here, since it's only called directly by Python space that already sets all the defaults (and only _intended_ to be called that way), I'm not sure that the `kwargs` indirection is achieving anything over just making all the configuration options explicitly required.

jakelishman commented on crates/qasm3/src/lib.rs:
```diff
@@ -150,11 +154,147 @@ pub fn load(
+    stream.call_method1("write", (output_str,))?;
+
+    Ok(())
+}
```
Accepting a text-mode stream in the Python-space `dump` isn't going to transition well to Rust - if we're given an open Python-space stream, we almost certainly should handle it by creating a full string from Rust and then copying it across to Python space, unless this is a performance pipeline we _really_ want to optimise, in which case we should make a Rust object that implements `Write`, and buffers writes out to the underlying Python stream so we aren't paying a huge runtime penalty for going Rust -> Python all the time (like how general file-backed streams also buffer writes).

For the time being, I feel like the whole Rust-space `dump` is unnecessary, and we could achieve the same thing just in the Python-space wrapper, with it calling Rust-space `dumps`.

jakelishman commented on qiskit/qasm3/__init__.py:
```diff
@@ -370,3 +369,76 @@ def load_experimental(pathlike_or_filelike, /, *, custom_gates=None, include_pat
+        includes = ["stdgates.inc"]
+    if basis_gates is None:
+        basis_gates = ["U"]
+    return _qasm3.dumps(
+        circuit,
+        {
+            "includes": includes,
+            "basis_gates": basis_gates,
+            "disable_constants": disable_constants,
+            "alias_classical_registers": alias_classical_registers,
+            "allow_aliasing": allow_aliasing,
+            "indent": indent,
+        },
+    )
```
This is the strange calling convention I mentioned in the previous review comment - it'd be more natural if these were just regular keyword arguments, even if it is just an internal function.

jakelishman reviewed (approved):
Thanks for the work on this Kento, and the updates from the last review.  In general, given this is in the "experimental" interface, and I've been sat on it a long time, I want to move to merge as-is, based solely on the observable user behaviour.  That said, I think there's a few places in the code where I'd look to improvements, for any of runtime performance, ease of future maintenance, and code organisation.  I've left some comments to these effects, which we might want to convert into full issues to track potential follow-up work (alongside obviously making it feature complete).

I fixed up the last merge conflicts, including some parsing of the Python-space kwargs that I'm not 100% sure how it worked before, but I assume is something that changed in a PyO3 update.

---

About `Identifier` vs interning: there should be no difference in safety, because as long as every `Interned<str>` refers to keys in the same `Interner`, equality comparisons between strings and interned keys are the same thing, just comparisons between interned keys (and the memory taken to store them) are much faster.  The trouble of comparing names between scopes is the same whether we use an explicit string or a key.  I'm happy for that to be a follow-up, though.

---

The review comments largely relate to the high-level API and the design of the data structures, and have some places we might want to change in the future in follow-ups as this drives towards being the primary exporter.  I didn't look at the actual implementation of the `QASM3Builder` methods in any significant detail - I know a lot of it is ported from the Python version (which comes with a lot of tech debt itself), and it's pretty dependent on the data structures, so if those change at all, a lot of the discussion of the individual methods would become obsolete.
